packages/float-input/src/useFloatInput.ts

```diff
@@ -0,0 +1,163 @@
+import numeral from 'numeral';
+import type {
+  AllHTMLAttributes,
+  FocusEventHandler,
+  FormEventHandler,
+} from 'react';
+import { useCallback, useState } from 'react';
+
+/** Ensure only valid values are passed to the consumer */
+const isValid = (text: string) => {
+  let valid = /^\s*\$?-?[\d,]+(\.\d+)?\s*$/;
+  return valid.test(text) && !isNaN(parse(text));
+};
+
+/** Strip invalid characters */
+const normalise = (text: string) => {
+  return text.trim().replace(/^\$/, '').replace(/,/g, '');
+};
+
+/** Safe parse */
+const parse = (text: string) => {
+  return parseFloat(normalise(text));
+};
+
+/** Resolve leading decimal issues */
+const padLeadingDecimal = (text: string) => {
+  if (text.startsWith('.')) {
+    return `0${text}`;
+  } else if (text.startsWith('-.')) {
+    return `-0${text.substring(1)}`;
+  }
+
+  return text;
+};
+
+/** Prepare the formatter */
+const useFormat = (fractionDigits?: number) => {
+  return useCallback(
+    (val?: number | string) => {
+      if (typeof val === 'undefined') {
+        return '';
+      }
+
+      if (typeof val === 'string' && !isValid(val)) {
+        return val;
+      }
+
+      const pattern = fractionDigits
+        ? `0,0.${'0'.repeat(fractionDigits)}`
+        : '0,0.[00000000]';
+      return numeral(val).format(pattern);
+    },
+    [fractionDigits]
+  );
+};
+
+type FloatInputHandlers = {
+  onChange: FormEventHandler<HTMLInputElement>;
+  onBlur: FocusEventHandler<HTMLInputElement>;
+  onFocus: FocusEventHandler<HTMLInputElement>;
+  value: string | number;
+};
+
+type UseFloatInputControlledProps = {
+  fractionDigits?: number;
+  onChange: (value: number | string) => void;
+  value: number | string;
+} & Pick<AllHTMLAttributes<HTMLInputElement>, 'onFocus' | 'onBlur'>;
+
+type UseFloatInputUncontrolledProps = {
+  fractionDigits?: number;
+  onChange?: never;
+  value?: never;
+} & Pick<AllHTMLAttributes<HTMLInputElement>, 'onFocus' | 'onBlur'>;
+
+export type UseFloatInputProps =
+  | UseFloatInputControlledProps
+  | UseFloatInputUncontrolledProps;
```

Comment:
Theres a bit of double-up on the discriminated union here. Maybe something like this?

```suggestion
type ControlledFloatInputProps = {
  onChange: (value: number | string) => void;
  value: number | string;
};

type UncontrolledFloatInputProps = {
  onChange?: never;
  value?: never;
};

export type UseFloatInputProps = {
  fractionDigits?: number;
  onFocus?: AllHTMLAttributes<HTMLInputElement>['onFocus'];
  onBlur?: AllHTMLAttributes<HTMLInputElement>['onBlur'];
} & (ControlledFloatInputProps | UncontrolledFloatInputProps);
```